neps/nep-0452.md

```diff
@@ -0,0 +1,234 @@
+- the user will be able to sign transactions on behalf of the new account using `privKey2`.
+
+#### Claiming a linkdrop with a NEAR Account
+
+A user with an _existing_ wallet that is claiming the assets with an existing public key, already registered on the linkdrop contract:
+
+- calls `claim`. The transaction is signed on behalf of the linkdrop contract (`signer_id` is set to the contract address) using `privKey1`.
+  - the args of this function call will simply contain the user's existing account ID.
+  - the linkdrop contract will delete the access key associated with `pubKey1` so that it cannot be used again.
+  - the linkdrop contract will transfer the funds to that account alongside any other assets.
+
+```ts
+/// Information about a specific public key.
+type KeyInfo = {
+   /// How much Gas should be attached when the key is used to call `claim` or `create_account_and_claim`.
+   /// It is up to the smart contract developer to calculate the required gas (which can be done either automatically on the contract or on the client-side).
+   required_gas: string,
+
+   /// yoctoNEAR$ amount that will be sent to the account that claims the linkdrop (either new or existing)
+   /// when the key is successfully used.
+   yoctoNEAR: string,
+
+   /// If using the NFT standard extension, a set of NFTData can be linked to the public key
+   /// indicating that all those assets will be sent to the account that claims the linkdrop (either new or
+   /// existing) when the key is successfully used.
+   nft_data: NFTData[] | null,
+
+   /// If using the FT standard extension, a set of FTData can be linked to the public key
+   /// indicating that all those assets will be sent to the account that claims the linkdrop (either new or
+   /// existing) when the key is successfully used.
+   ft_data: FTData[] | null
+
+   /// ... other types can be introduced and the standard is easily extendable.
+}
+
+
+/// Data outlining a specific Non-Fungible Token that should be sent to the claiming account
+/// (either new or existing) when a key is successfully used.
+type NFTData = {
+   /// the id of the token to transfer
+   token_id: string,
+
+   /// The valid NEAR account indicating the Non-Fungible Token contract.
+   contract_id: string
+}
+
+
+/// Data outlining Fungible Tokens that should be sent to the claiming account
+/// (either new or existing) when a key is successfully used.
+type FTData = {
+   /// The number of tokens to transfer, wrapped in quotes and treated
+   /// like a string, although the number will be stored as an unsigned integer
+   /// with 128 bits.
+   amount: string,
+
+   /// The valid NEAR account indicating the Fungible Token contract.
+   contract_id: string
+}
+
+
+
+
+/****************/
+/* VIEW METHODS */
+/****************/
+
+/// Returns the $yoctoNEAR amount associated with a given public key
+/// Panics if the key does not exist.
```

Comment:
The reason for the panic is to dictate whether or not the linkdrop actually exists. If you were to `simply return 0`, wallets wouldn't know when linkdrops are valid or not. 

The only other solution could be to make the return value optional `Option<KeyInfo>`. I chose to make it panic as that's what pretty much every wallet assumes at the moment (original linkdrop contract does this).

Reply:
If the key doesn't exist it **absolutely** should panic. This is what the original linkdrop standard does, and we should remain consistent otherwise existing implementations will break. What reason do you have to return 0 balance for a key that doesn't exist on the contract?

Reply:
Does your DB panic if you query a non existing row?  The problem with panics is that you can't manage them easily. However you can easily manage return types.

Reply:
IMHO, panics in queries should be avoided whenever possible.

"HTTP Error 404" is not a panic. Also, HTTP has 4xx errors for user request errors and 5xx for internal errors. Let me make few more examples:
* querying not existing balance in NEAR FT, NFT doesn't panic. Nor ERC-20, nor ERC-1155....
* if you look at the openzepplin implementation, most (if not all) queries are not throwing exceptions
* imo, DB is a good example (decades of polishing the patterns)

> but with the current state of things, panics are conventionally used to return invariant errors.

Depends how do you look at it. Error state is something that you can't handle. In good software engineering panics and exceptions, whenever possible, should be deferred to the top of the call stack, rather than poisoning call stack.
In Ethereum, they seem to understand that with the evolution of the native transfers. The current recommendation ([1](https://solidity-by-example.org/sending-ether/), [2](https://ethereum.stackexchange.com/questions/19341/address-send-vs-address-transfer-best-practice-usage)) is to use `call` (which return bool) rather than `transfer` (which panics).

> If you were to simply return 0, wallets wouldn't know when linkdrops are valid or not.

@BenKurrek, creating a 0 linkdrop is a wrong transaction -> it should be rejected when the linkdrop is created.
Also, even if there is a 0 linkdrop... how a wallet should handle it? What's the point of having it, rather than filtering out? With tokens, you don't show 0 balances.

Reply:
@robert-zaremba with Zero-balance accounts now there can be meaningful linkdrops with 0 NEAR and some FT/NFT.

Reply:
@frol  Didn't we [decide](https://github.com/near/NEPs/pull/452/files/845cd53a080c7688a6a4583047980ab29640c7e0#r1147491782) that we are removing `get_key_balance` and use `get_key_information` instead?

I think this thread is more about `get_key_information` which returns struct, where all fields can be zero.

Reply:
@robert-zaremba creating a linkdrop with 0 balance is definitely *not* a wrong transaction. There are many applications to this (transferring NFTs, FTs, function calls etc.) This is further reinforced with the introduction of zero balance accounts.